### PR TITLE
Upstream PR for all bugfixes since first commit of Nutanix Prism checks

### PR DESCRIPTION
**Pull request by Yogibaer75**
This PR request includes two bug fixes and 1 missing check for Nutanix.

Missing checks - host network interfaces
Bug fixes 
- vm memory -> unknown if no memory values existing fixed
- vm_tools -> state can be ignored

### CODE REVIEW

**Comment by Yogibaer75** [comment]
I have read the CLA Document and I hereby sign the CLA or my organization already has a signed CLA.

**Comment by martinhv** [comment]
From a PM perspective, I approve this change. It creates value for Checkmk and fixes bugs. 
The technical evaluation I leave to dev.

File: cmk/base/plugins/agent_based/prism_host_networks.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        params,
+        _create_interface(section),
+        value_store=value_store,
+    )
+
+
+def check_prism_host_networks(
+    item: str, params: Mapping[str, Any], section: Section
+) -> CheckResult:
+    data = section.get(item)
```
Comment (BenediktSeidl) [inline cmk/base/plugins/agent_based/prism_host_networks.py:100]: are you sure that those statements are really needed?
i would think that this case is already handled by `interfaces.check_multiple_interfaces`.

**Review by BenediktSeidl — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. In general, it looks good, it even has tests :heart: very nice!

To merge I would add three (or two) werks, but I need some more context to the following items:
* vm memory -> unknown if no memory values existing fixed
* vm_tools -> state can be ignored
It would be good to know what the previous behavior was, and how it was fixed, so I can come up with a Werk that other customers can understand.

Can you split your two commits into three commits? Two commits for the two fixes and an additional commit for the new feature.

You can add the style change in `cmk/base/plugins/agent_based/prism_cluster_io.py` to any of these commits, but it would be cool if you could mention it in the commit message.

This probably seems overly picky, but git history is often the only way to understand the intent of changes from years ago, so we try to maintain it as best we can.

I've also left four code annotations with questions and comments, I'm open to discussiong these issues.

File: cmk/base/plugins/agent_based/prism_host_networks.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    return parsed
+
+
+register.agent_section(
+    name="prism_host_networks",
+    parse_function=parse_prism_host_networks,
+)
+
+
+def _create_interface(section: Section) -> interfaces.Section[interfaces.InterfaceWithRates]:
```
[the remark this answers is not in the document]
Comment (Yogibaer75) [reply cmk/base/plugins/agent_based/prism_host_networks.py:39]: I must added this workaround to make tests working. I copied this part from one of your own checks :)

File: cmk/base/plugins/agent_based/prism_host_networks.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+# -*- encoding: utf-8; py-indent-offset: 4 -*-
```
[the remark this answers is not in the document]
Comment (Yogibaer75) [reply cmk/base/plugins/agent_based/prism_host_networks.py:2]: Why is this now a problem in all the other checks made before there is the same header?

File: cmk/base/plugins/agent_based/utils/prism.py [hunk 4]
```diff
@@ -39,3 +40,10 @@ def load_json(string_table: StringTable) -> PrismAPIData:
         return json.loads(string_table[0][0])
     except (IndexError, json.decoder.JSONDecodeError):
         return {}
+
+
+def load_json2(string_table: StringTable) -> PrismAPIData2:
```
[the remark this answers is not in the document]
Comment (Yogibaer75) [reply cmk/base/plugins/agent_based/utils/prism.py:45]: PrismApiData --> dict[str, Any]
PrismApiData2 --> dict[dict, Any]
Again this extra code was only made to make the testing and checks happy.

**Comment by Yogibaer75** [comment]
> * vm memory -> unknown if no memory values existing fixed
It is possible the one defined VM don't report a memory value and then the check was crashing if no value exists.

> * vm_tools -> state can be ignored
>   It would be good to know what the previous behavior was, and how it was fixed, so I can come up with a Werk that other customers can understand.
Before you needed to specify a wanted state. If nothing was specified the default parameter state was used. Now you can define to ignore the state what can be helpful for machines where no tools can be installed like appliances and so on.

> Can you split your two commits into three commits? Two commits for the two fixes and an additional commit for the new feature.
After the submit no, I don't know how to split the commits.

> You can add the style change in `cmk/base/plugins/agent_based/prism_cluster_io.py` to any of these commits, but it would be cool if you could mention it in the commit message.

Ok that's a point. As this was a commit of all the collected changes from my repo it is possible that i forgot to mention.

> This probably seems overly picky, but git history is often the only way to understand the intent of changes from years ago, so we try to maintain it as best we can.
> 
> I've also left four code annotations with questions and comments, I'm open to discussiong these issues.
I wrote some comments. All these "problems" are only there as your tests wanted these changes. Before my simple code was rejected from the mypy/pytest tests 🤣 

I think some constructs are over complicated inside CMK if you want to make the test environment happy.
Later bugfixes a can submit as simple small PRs, but then we also need a better way for the processing of these PRs.

**Comment by BenediktSeidl** [comment]
Thanks for your response and the additional info about your change. We are happy that you share your domain knowledge with checkmk and want to include this in our product.
I would suggest, that I take over from here, adapt your code to meet our coding standards, split it into three commits and add the necessary werks. The business logic will not be altered. You are still the main author of the git commits, to show that there were additional changes, I would add myself as a co-author.
Is this okay for you?

File: cmk/base/plugins/agent_based/prism_host_networks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenediktSeidl) [reply cmk/base/plugins/agent_based/prism_host_networks.py:2]: before asking you to change this, i asked internally and the official statement is, that we should only use the checkmk license header.
your files were the only exception that used non-official license headers. our license header check was very lax in this regard. today i merged a change which applies stricter rules from now on.

**Comment by Yogibaer75** [comment]
> Is this okay for you?

Yes why not :)

File: cmk/base/plugins/agent_based/utils/prism.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenediktSeidl) [reply cmk/base/plugins/agent_based/utils/prism.py:45]: ah! now i see the difference, sorry about this.

**Comment by BenediktSeidl** [comment]
@Yogibaer75 yesterday i merged a first version of your change into the master branch. can you please check in a daily master build if it is working as intended? i've implemented some bigger changes for parsing the json file, as it was suggested in the internal code review.

the internal code review also opened the question if the api already returns rates, or if those are global counters? as far as i understand those are rates based on a 30 seconds interval. is the comment i added correct? https://github.com/Checkmk/checkmk/blob/b9f640b864f2b284d0bb33efbf095471929da339/cmk/base/plugins/agent_based/prism_host_networks.py#L78-L80 

thanks again for your contribution!

**Comment by Yogibaer75** [comment]
> the internal code review also opened the question if the api already returns rates, or if those are global counters? as far as i understand those are rates based on a 30 seconds interval. is the comment i added correct?
> 
Yes these are already rates and no counters. And the rate is based on the last 30 seconds. I found no other way than this to get the interface data.

For the master build i must check the next days if it does what it should :)

**Pull request closed without merging**